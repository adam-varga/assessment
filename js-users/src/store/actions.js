// ACTION TYPES

export const GET_USERS__REQUEST = "GET_USERS__REQUEST";
export const GET_USERS__SUCCEEDED = "GET_USERS__SUCCEEDED";
export const GET_USERS__FAILED = "GET_USERS__FAILED";
export const GET_USERS__RESET = "GET_USERS__RESET";

export const CREATE_USER__REQUEST = "CREATE_USER__REQUEST";
export const CREATE_USER__SUCCEEDED = "CREATE_USER__SUCCEEDED";
export const CREATE_USER__FAILED = "CREATE_USER__FAILED";
export const CREATE_USER__RESET = "CREATE_USER__RESET";


export const UPDATE_USER__REQUEST = "UPDATE_USER__REQUEST";
export const UPDATE_USER__SUCCEEDED = "UPDATE_USER__SUCCEEDED";
export const UPDATE_USER__FAILED = "UPDATE_USER__FAILED";
export const UPDATE_USER__RESET = "UPDATE_USER__RESET";


// ACTION CREATORS

export function createAction(type, payload) {
  return {
    type,
    payload,
  };
}
