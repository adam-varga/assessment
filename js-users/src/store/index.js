import { createStore, applyMiddleware } from "redux";
import reducer from "./reducers";
import rootSaga from "./sagas";
import { composeWithDevTools } from "redux-devtools-extension";
import createSagaMiddleware from "redux-saga";

export default function configureStore() {
  const sagaMiddleware = createSagaMiddleware();

  const store = createStore(
    reducer,
    composeWithDevTools(applyMiddleware(sagaMiddleware))
  );

  sagaMiddleware.run(rootSaga);

  return store;
}
