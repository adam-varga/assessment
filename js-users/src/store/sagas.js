import { put, takeEvery, all } from "redux-saga/effects";
import {
  createAction,
  GET_USERS__REQUEST,
  GET_USERS__SUCCEEDED,
  GET_USERS__FAILED,
} from "./actions";

function* getUsers() {
  yield takeEvery(GET_USERS__REQUEST, function* () {
    try {
      let users = yield fetch(
        "http://js-assessment-backend.herokuapp.com/users"
      );

      users = yield users.json();

      const usersById = {};
      const userIds = [];

      users.forEach((user) => {
        usersById[user.id] = user;
        userIds.push(`${user.id}`);
      });

      yield put(
        createAction(GET_USERS__SUCCEEDED, {
          usersById,
          userIds,
        })
      );
    } catch (error) {
      yield put(createAction(GET_USERS__FAILED, { error }));
    }
  });
}

export default function* rootSaga() {
  yield all([getUsers()]);
}
