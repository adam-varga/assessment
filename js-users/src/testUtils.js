import { createStore } from "redux";
import React from "react";
import { Provider } from "react-redux";
import { render } from "@testing-library/react";
import reducer from "./store/reducers";

export function renderWithRedux(
  ui,
  { initialState, store = createStore(reducer, initialState) } = {}
) {
  return {
    ...render(<Provider store={store}>{ui}</Provider>),
    store,
  };
}
