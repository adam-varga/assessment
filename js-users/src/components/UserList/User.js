import React from "react";
import PropTypes from "prop-types";
import {
  Item,
  Info,
  ToggleUserStatusButton,
  Actions,
  EditButton,
  Name,
  CreatedAt,
} from "./UserList.styles";
import { connect } from "react-redux";
import { createAction, UPDATE_USER__REQUEST } from "../../store/actions";
import { USER_STATUSES, REQUEST_STATUSES } from "../../constants";
import { FiUnlock, FiLock, FiEdit2 } from "react-icons/fi";
import get from "lodash.get";

const statusButtonIconMap = {
  [USER_STATUSES.LOCKED]: FiLock,
  [USER_STATUSES.ACTIVE]: FiUnlock,
};

class User extends React.Component {
  render() {
    const { first_name, last_name, created_at, status, id } = this.props.user;

    const StatusButtonIcon = statusButtonIconMap[status];

    return (
      <Item>
        <Info {...this.props.user}>
          <Name>{`${first_name} ${last_name}`}</Name>
          <CreatedAt>{new Date(created_at).toLocaleDateString()}</CreatedAt>
        </Info>
        <Actions>
          <ToggleUserStatusButton
            data-testid={`toggle-user-status-button_${id}`}
            status={status}
            active={status === USER_STATUSES.LOCKED}
            onClick={this.props.toggleUserStatus}
            disabled={this.props.requestStatus === REQUEST_STATUSES.PENDING}
          >
            <StatusButtonIcon />
          </ToggleUserStatusButton>
          <EditButton
            to={`/edit/${id}`}
            data-testid={`edit-user-button_${id}`}
            disabled={this.props.requestStatus === REQUEST_STATUSES.PENDING}
          >
            <FiEdit2 />
          </EditButton>
        </Actions>
      </Item>
    );
  }
}

User.propTypes = {
  user: PropTypes.shape({
    first_name: PropTypes.string.isRequired,
    last_name: PropTypes.string.isRequired,
    status: PropTypes.oneOf(Object.values(USER_STATUSES)).isRequired,
  }).isRequired,
  requestStatus: PropTypes.oneOf(Object.values(REQUEST_STATUSES)),
  toggleUserStatus: PropTypes.func.isRequired,
};

function mapStateToProps(state) {
  return {
    requestStatus: get(state, `requests.${UPDATE_USER__REQUEST}.status`),
  };
}

function mapDispatchToProps(dispatch, ownProps) {
  return {
    toggleUserStatus: () =>
      dispatch(
        createAction(UPDATE_USER__REQUEST, {
          ...ownProps.user,
          status:
            ownProps.user.status === USER_STATUSES.LOCKED
              ? USER_STATUSES.ACTIVE
              : USER_STATUSES.LOCKED,
        })
      ),
  };
}

export default connect(mapStateToProps, mapDispatchToProps)(User);
