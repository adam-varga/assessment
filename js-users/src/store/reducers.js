import { GET_USERS__SUCCEEDED } from "./actions";
import { combineReducers } from "redux";

// USERS

function users(state = {}, action) {
  switch (action.type) {
    case GET_USERS__SUCCEEDED: {
      const { usersById, userIds } = action.payload;

      return {
        usersById,
        userIds,
      };
    }
    default:
      return state;
  }
}

// REQUESTS

const REQUEST_REGEXPS = {
  START: /__REQUEST$/,
  SUCCESS: /__SUCCEEDED$/,
  FAILURE: /__FAILED$/,
};

function getRequestActionType(actionType, regExp) {
  return `${actionType.substring(0, actionType.match(regExp).index)}__REQUEST`;
}

function requests(state = {}, action) {
  if (action.type.match(REQUEST_REGEXPS.START)) {
    return {
      ...state,
      [action.type]: {
        loading: true,
      },
    };
  }

  if (action.type.match(REQUEST_REGEXPS.SUCCESS)) {
    return {
      ...state,
      [getRequestActionType(action.type, REQUEST_REGEXPS.SUCCESS)]: {
        loading: false,
      },
    };
  }

  if (action.type.match(REQUEST_REGEXPS.FAILURE)) {
    return {
      ...state,
      [getRequestActionType(action.type, REQUEST_REGEXPS.FAILURE)]: {
        loading: false,
        error: action.payload.error,
      },
    };
  }

  return state;
}

export default combineReducers({
  users,
  requests,
});
