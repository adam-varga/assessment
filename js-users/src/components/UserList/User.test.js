import React from "react";
import { cleanup, fireEvent } from "@testing-library/react";
import User from "./User";
import { renderWithRedux } from "../../testUtils";
import { createAction, UPDATE_USER__REQUEST } from "../../store/actions";
import { createStore } from "redux";
import reducer from "../../store/reducers";
import { BrowserRouter as Router } from "react-router-dom";

afterEach(cleanup);

describe("User", () => {
  let renderResult;
  let store;
  const userProps = {
    id: 1,
    last_name: "Person",
    first_name: "Test",
    status: "locked",
    created_at: "2014-01-10T08:28:49.030Z",
    updated_at: "2020-06-12T14:36:40.351Z",
    url: "http://1.json",
  };

  beforeEach(() => {
    store = createStore(reducer);

    store.dispatch = jest.fn();

    renderResult = renderWithRedux(
      <Router>
        <User user={userProps} />
      </Router>,
      { store }
    );
  });

  it("renders correctly", () => {
    const { getByText, asFragment } = renderResult;

    expect(getByText("Test Person")).toBeInTheDocument();
    expect(
      getByText(new Date(userProps.created_at).toLocaleDateString())
    ).toBeInTheDocument();
    expect(asFragment()).toMatchSnapshot();
  });

  it("dispatches UPDATE_USER__REQUEST on status toggle button click", () => {
    const { getByTestId, store } = renderResult;
    const statusToggleButton = getByTestId(
      `toggle-user-status-button_${userProps.id}`
    );

    fireEvent(
      statusToggleButton,
      new MouseEvent("click", {
        bubbles: true,
        cancelable: true,
      })
    );

    expect(store.dispatch).toHaveBeenCalledWith(
      createAction(UPDATE_USER__REQUEST, {
        ...userProps,
        status: "active",
      })
    );
  });

  it("changes URL after clicking Edit button", () => {
    const { getByTestId } = renderResult;
    const editButton = getByTestId(`edit-user-button_${userProps.id}`);

    expect(global.location.pathname).toBe("/");

    fireEvent(
      editButton,
      new MouseEvent("click", {
        bubbles: true,
        cancelable: true,
      })
    );

    expect(global.location.pathname).toBe(`/edit/${userProps.id}`);
  });
});
