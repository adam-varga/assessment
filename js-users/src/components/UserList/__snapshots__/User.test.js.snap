// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`User renders correctly 1`] = `
<DocumentFragment>
  <li
    class="sc-fzozJi jEWWTR"
  >
    <div
      class="sc-AxhCb bWMbKK"
      id="1"
    >
      <p
        class="sc-AxhUy bThcWD"
      >
        Test Person
      </p>
      <p
        class="sc-AxgMl eyrKYH"
      >
        1/10/2014
      </p>
    </div>
    <div
      class="sc-AxheI kSIgOR"
    >
      <button
        class="sc-AxjAm sc-AxmLO bgXSKk"
        data-testid="toggle-user-status-button_1"
      >
        <svg
          fill="none"
          height="1em"
          stroke="currentColor"
          stroke-linecap="round"
          stroke-linejoin="round"
          stroke-width="2"
          viewBox="0 0 24 24"
          width="1em"
          xmlns="http://www.w3.org/2000/svg"
        >
          <rect
            height="11"
            rx="2"
            ry="2"
            width="18"
            x="3"
            y="11"
          />
          <path
            d="M7 11V7a5 5 0 0 1 10 0v4"
          />
        </svg>
      </button>
      <a
        class="sc-AxirZ sc-Axmtr eQgWKs"
        data-testid="edit-user-button_1"
        href="/edit/1"
      >
        <svg
          fill="none"
          height="1em"
          stroke="currentColor"
          stroke-linecap="round"
          stroke-linejoin="round"
          stroke-width="2"
          viewBox="0 0 24 24"
          width="1em"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M17 3a2.828 2.828 0 1 1 4 4L7.5 20.5 2 22l1.5-5.5L17 3z"
          />
        </svg>
      </a>
    </div>
  </li>
</DocumentFragment>
`;
